File: Xamarin/SampleXamarin.Android/AnchorCreationFragment.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using Android.OS;
using Android.Support.V4.App;
using Android.Views;
using Android.Widget;
using Microsoft.Azure.SpatialAnchors;
using Xamarin.Essentials;
using static Android.Views.View;

namespace SampleXamarin
{
    internal class AnchorCreationFragment : Fragment, IOnClickListener
    {
        private bool isCreatingAnchor = false;

        private Button createAnchorButton;
        private ProgressBar requiredScanProgress;
        private ProgressBar recommendedScanProgress;

        public delegate void AnchorCreatedListener(AnchorVisual createdAnchor);
        public delegate void AnchorCreationFailedListener(AnchorVisual placedAnchor, string errorMessage);

        public AzureSpatialAnchorsManager CloudAnchorManager { private get; set; }
        public AnchorVisual PlacedVisual { private get; set; }
        public AnchorCreatedListener OnAnchorCreated { private get; set; }
        public AnchorCreationFailedListener OnAnchorCreationFailed { private get; set; }

        public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
        {
            return inflater.Inflate(Resource.Layout.coarse_reloc_anchor_creation, container, false);
        }

        public override void OnViewCreated(View view, Bundle savedInstanceState)
        {
            createAnchorButton = view.FindViewById<Button>(Resource.Id.create_anchor);
            requiredScanProgress = view.FindViewById<ProgressBar>(Resource.Id.required_scan_progress);
            recommendedScanProgress = view.FindViewById<ProgressBar>(Resource.Id.recommended_scan_progress);
        }

        public override void OnCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            if (CloudAnchorManager == null)
            {
                FragmentHelper.BackToPreviousFragment(Activity);
            }
        }

        public override void OnStart()
        {
            base.OnStart();

            createAnchorButton.Enabled = false;
            createAnchorButton.SetOnClickListener(this);
            CloudAnchorManager.OnSessionUpdated += OnSessionUpdated;
        }

        public override void OnStop()
        {
            if (CloudAnchorManager != null)
            {
                CloudAnchorManager.OnSessionUpdated -= OnSessionUpdated;
            }

            if (PlacedVisual != null)
            {
                PlacedVisual.Destroy();
                PlacedVisual = null;
            }

            base.OnStop();
        }

        private bool CanCreateAnchor()
        {
            return !isCreatingAnchor && PlacedVisual != null;
        }

        private void OnSessionUpdated(object sender, SessionUpdatedEvent sessionUpdatedEvent)
        {
            float requiredForCreateProgress = sessionUpdatedEvent.Status.ReadyForCreateProgress;
            float recommendedForCreateProgress = sessionUpdatedEvent.Status.RecommendedForCreateProgress;

            MainThread.BeginInvokeOnMainThread(() =>
            {
                requiredScanProgress.Progress = (int)(100 * requiredForCreateProgress);
                recommendedScanProgress.Progress = (int)(100 * recommendedForCreateProgress);
                bool allowCreation = CanCreateAnchor() && requiredForCreateProgress >= 1.0f;
                createAnchorButton.Enabled = allowCreation;
            });
        }

        async void IOnClickListener.OnClick(View view)
        {
            if (!CanCreateAnchor())
            {
                return;
            }

            createAnchorButton.Enabled = false;
            CloudSpatialAnchor cloudAnchor = new CloudSpatialAnchor();
            cloudAnchor.LocalAnchor = PlacedVisual.LocalAnchor;
            cloudAnchor.AppProperties.Add("Shape", PlacedVisual.Shape.ToString());

            isCreatingAnchor = true;

            await CloudAnchorManager.CreateAnchorAsync(cloudAnchor);
            if ((cloudAnchor.Identifier?.Length ?? 0) == 0)
            {
                OnAnchorCreationFailed?.Invoke(PlacedVisual, "Failed to create anchor");
                return;
            }

            AnchorVisual createdAnchor = PlacedVisual;
            PlacedVisual = null;
            OnAnchorCreated?.Invoke(createdAnchor);
        }
    }
}

File: Android/Java/app/src/main/java/com/microsoft/sampleandroid/AnchorCreationFragment.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
package com.microsoft.sampleandroid;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;

import com.microsoft.azure.spatialanchors.CloudSpatialAnchor;
import com.microsoft.azure.spatialanchors.CloudSpatialException;
import com.microsoft.azure.spatialanchors.SessionUpdatedEvent;

public class AnchorCreationFragment extends Fragment {
    private AzureSpatialAnchorsManager cloudAnchorManager;
    private AnchorCreationListener listener;
    private boolean isCreatingAnchor = false;
    private AnchorVisual placedVisual;

    private Button createAnchorButton;
    private ProgressBar requiredScanProgress;
    private ProgressBar recommendedScanProgress;

    public void setCloudAnchorManager(AzureSpatialAnchorsManager cloudAnchorManager) {
        this.cloudAnchorManager = cloudAnchorManager;
    }

    public void setPlacement(AnchorVisual placedVisual) {
        this.placedVisual = placedVisual;
    }

    public void setListener(AnchorCreationListener listener) {
        this.listener = listener;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.coarse_reloc_anchor_creation, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        createAnchorButton = (Button)view.findViewById(R.id.create_anchor);
        requiredScanProgress = (ProgressBar)view.findViewById(R.id.required_scan_progress);
        recommendedScanProgress = (ProgressBar)view.findViewById(R.id.recommended_scan_progress);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (cloudAnchorManager == null) {
            FragmentHelper.backToPreviousFragment(getActivity());
        }
    }

    @Override
    public void onStart() {
        super.onStart();

        createAnchorButton.setEnabled(false);
        createAnchorButton.setOnClickListener(this::onCreateAnchorClicked);
        cloudAnchorManager.addSessionUpdatedListener(this::onSessionUpdated);
    }

    @Override
    public void onStop() {
        if (cloudAnchorManager != null) {
            cloudAnchorManager.removeSessionUpdatedListener(this::onSessionUpdated);
        }

        if (placedVisual != null) {
            placedVisual.destroy();
            placedVisual = null;
        }

        super.onStop();
    }

    private boolean canCreateAnchor() {
        return !isCreatingAnchor && placedVisual != null;
    }

    private void onSessionUpdated(SessionUpdatedEvent sessionUpdatedEvent) {
        float requiredForCreateProgress = sessionUpdatedEvent.getStatus().getReadyForCreateProgress();
        float recommendedForCreateProgress = sessionUpdatedEvent.getStatus().getRecommendedForCreateProgress();
        MainThreadContext.runOnUiThread(() -> {
            requiredScanProgress.setProgress((int)(100 * requiredForCreateProgress));
            recommendedScanProgress.setProgress((int)(100 * recommendedForCreateProgress));
            boolean allowCreation = canCreateAnchor() && requiredForCreateProgress >= 1.0f;
            createAnchorButton.setEnabled(allowCreation);
        });
    }

    private void onCreateAnchorClicked(View view) {
        if (!canCreateAnchor()) {
            return;
        }

        createAnchorButton.setEnabled(false);
        CloudSpatialAnchor cloudAnchor = new CloudSpatialAnchor();
        cloudAnchor.setLocalAnchor(placedVisual.getLocalAnchor());
        cloudAnchor.getAppProperties().put("Shape", placedVisual.getShape().toString());

        isCreatingAnchor = true;
        cloudAnchorManager
            .createAnchorAsync(cloudAnchor)
            .whenComplete((anchor, thrown) -> {
                MainThreadContext.runOnUiThread(() -> {
                    isCreatingAnchor = false;
                    if (placedVisual == null) {
                        return;
                    }

                    if (thrown != null) {
                        if (listener != null) {
                            String errorMessage = getErrorMessageFromThrowable(thrown);
                            listener.onAnchorCreationFailed(placedVisual, errorMessage);
                        }
                        return;
                    }

                    AnchorVisual createdAnchor = placedVisual;
                    placedVisual = null;
                    if (listener != null) {
                        listener.onAnchorCreated(createdAnchor);
                    }
                });
            });
    }

    private String getErrorMessageFromThrowable(Throwable thrown) {
        Throwable originalException = thrown;
        while (originalException != null && !(originalException instanceof CloudSpatialException)) {
            originalException = originalException.getCause();
        }
        if (originalException != null) {
            CloudSpatialException azureSpatialAnchorsException = (CloudSpatialException) originalException;
            return azureSpatialAnchorsException.getErrorCode().toString();
        } else {
            return thrown.toString();
        }
    }
}
